core/src/main/java/org/jdbi/v3/Handle.java

```diff
@@ -28,109 +32,231 @@
 import org.jdbi.v3.tweak.ColumnMapper;
 import org.jdbi.v3.tweak.RowMapper;
 import org.jdbi.v3.tweak.StatementBuilder;
+import org.jdbi.v3.tweak.StatementCustomizer;
 import org.jdbi.v3.tweak.StatementLocator;
+import org.jdbi.v3.tweak.TransactionHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * This represents a connection to the database system. It usually is a wrapper around
  * a JDBC Connection object.
  */
-public interface Handle extends Closeable
+public class Handle implements Closeable
 {
+    private static final Logger LOG = LoggerFactory.getLogger(Handle.class);
 
-    /**
-     * @return the JDBC Connection this Handle uses
-     */
-    Connection getConnection();
+    protected final JdbiConfig config;
+    protected StatementBuilder statementBuilder;
+    private boolean closed = false;
+    protected final TransactionHandler transactions;
+    protected final Connection connection;
+
+    Handle(JdbiConfig config,
+            TransactionHandler transactions,
+            StatementBuilder preparedStatementCache,
+            Connection connection)
+    {
+        this.config = config;
+        this.statementBuilder = preparedStatementCache;
+        this.transactions = transactions;
+        this.connection = connection;
+    }
 
     /**
-     * Closes the handle, its connection, and any other database resources it is holding.
+     * Get the JDBC Connection this Handle uses.
      *
-     * @throws org.jdbi.v3.exceptions.UnableToCloseResourceException if any resources throw exception while closing
+     * @return the JDBC Connection this Handle uses
      */
-    @Override
-    void close();
+    public Connection getConnection() {
+        return this.connection;
+    }
 
     /**
      * Define a statement attribute which will be applied to all {@link StatementContext}
-     * instances for statements created from this handle
+     * instances for statements created from this handle.
      *
      * @param key Attribute name
      * @param value Attribute value
      */
-    void define(String key, Object value);
+    public Handle define(String key, Object value) {
+        config.statementAttributes.put(key, value);
+        return this;
+    }
+
+    public Handle registerArgumentFactory(ArgumentFactory argumentFactory) {
+        config.argumentRegistry.register(argumentFactory);
+        return this;
+    }
+
+    public Handle registerCollectorFactory(CollectorFactory factory) {
+        config.collectorRegistry.register(factory);
+        return this;
+    }
 
     /**
-     * Start a transaction
+     * Register a column mapper which will have its parameterized type inspected to determine what it maps to.
      *
-     * @return the same handle
+     * Column mappers may be reused by {@link RowMapper} to map individual columns.
+     *
+     * @param mapper the column mapper
```

Comment:
We should add a `@throws` here for the case that the mapper isn't concretely parameterized.